utils/vlog/vlog_to_pbtype.py

```diff
@@ -54,68 +54,225 @@
-    help="""\
-Command seperate list of include directories.
-""",
-    default=""
-)
-parser.add_argument('-o', help="""\
-Output filename, default 'model.xml'
-""")
 
-args = parser.parse_args()
-iname = os.path.basename(args.infiles[0])
+def make_timings(
+        pb_type_xml, sdf_file_name, sdf_cell_paths=None, sdf_variant="slow"
+):
+    """
+    Loads timings from a given SDF file and converts them to appropriate
+    statements within pb_type XML.
 
-outfile = "pb_type.xml"
-if "o" in args and args.o is not None:
-    outfile = args.o
+    If sdf_cell_name is None then the function tries to find in SDF cell with
+    name equal to the pb_type name.
 
-yosys.run.add_define("PB_TYPE")
-if args.includes:
-    for include in args.includes.split(','):
-        yosys.run.add_include(include)
+    If sdf_inst_name is None and there is only one instance of that cell in SDF
+    then that one is taken.
+    """
 
-vjson = yosys.run.vlog_to_json(args.infiles, flatten=False, aig=False)
-yj = YosysJSON(vjson)
+    def parse_timescale(timescale_str):
+        """
+        Parses a timescale expression and returns its numerical value
+        """
+        match = re.match("([0-9]+)(fs|ps|ns|us|ms|s?)$", timescale_str)
+
+        base = float(match.group(1))
+        suffix = match.group(2)
+
+        if suffix == "fs":
+            base *= 1e-15
+        elif suffix == "ps":
+            base *= 1e-12
+        elif suffix == "ns":
+            base *= 1e-9
+        elif suffix == "us":
+            base *= 1e-6
+        elif suffix == "ms":
+            base *= 1e-3
+
+        return base
+
+    def collect_timings(sdf, sdf_cell_paths):
+        """
+        Collects timings for multiple cells which paths are given in the path
+        list.
+        """
+
+        collected_timings = {}
+
+        # For each path
+        for sdf_cell_path in sdf_cell_paths:
+
+            # Only cell name given
+            if "/" not in sdf_cell_path:
+                sdf_cell_name = sdf_cell_path
+                sdf_inst_name = None
+            # Both cell name and instance path given
+            else:
+                sdf_inst_name, sdf_cell_name = sdf_cell_path.rsplit(
+                    "/", maxsplit=1
+                )
 
-if args.top is not None:
-    top = args.top
-else:
-    wm = re.match(r"([A-Za-z0-9_]+)\.sim\.v", iname)
-    if wm:
-        top = wm.group(1).upper()
-    else:
-        print(
-            "ERROR file name not of format %.sim.v ({}), cannot detect top level. Manually specify the top level module using --top"
-            .format(iname)
-        )
-        sys.exit(1)
+            # Get the SDF timing info for a particular CELL/INSTANCE
+            try:
+                sdf_cell = sdf["cells"][sdf_cell_name]
+            except KeyError:
+                print(
+                    "ERROR, the SDF file does not contain data for cell '{}'".
+                    format(sdf_cell_path)
+                )
+                continue
```

Comment:
why continue? If you want to skip the unexisting cells you can check for it with:
```
if sdf_cell_name in sdf["cells"]:
```

Reply:
Done.